### PR TITLE
Move to stable Rust: upgrade deps, enforce precision invariant, add error handling

### PR DESCRIPTION
**Pull request by danieleades**
Upgrade to arithmetic-coding 0.4, bitstream-io 2, syn 2, darling 0.20 and
thiserror 2, adapting the visitor to the 0.4 chained one-shot API. Delete the
`never_type`/`associated_type_defaults` nightly features (replacing `! ` with
`std::convert::Infallible`), add `#![forbid(unsafe_code)]`, and set MSRV 1.85
(the floor required by arithmetic-coding-core 0.4.2).

Soundness:
- Replace the array decode's `MaybeUninit::assume_init()` UB with a safe
  `std::array::from_fn` that captures a `&mut Result` and unwraps only after
  every element decodes successfully.

Error handling (issue #7):
- Add `DecodeError` (thiserror) with `Io` and `InvalidSymbol` variants; every
  decode path now returns `Result<_, DecodeError>` and never panics or unwraps
  on untrusted input. All `unreachable!()` arms become `InvalidSymbol`.
- Encode stays `io::Result`; `encode_bytes` uses `expect` with the invariant.

Precision invariant:
- Derive the largest u128-safe coder precision from upstream's arithmetic
  (`(2^P + 1) * denominator <= 2^128 - 1`): P = 64, so `MAX_DENOMINATOR = 2^62`.
  Expose `PRECISION`/`MAX_DENOMINATOR` consts with the derivation documented.
- Make `FloatModel::new` fallible (`ModelError`), rejecting NaN/inverted bounds
  and oversized denominators; validate float bounds at macro-expansion time in
  the derive and emit a spanned compile error.
- `OneShot` keeps `ValueError = Infallible` and debug-asserts symbols < N.

Also rename `Option__` to `OptionDiscriminant` (wire order None = 0, Some = 1),
migrate the derive to syn 2 / darling 0.20 (`attr.meta`, `path()`, a `Number`
wrapper restoring `min = -1.0` negative-literal support), fully-qualify the
generated trait calls, and implement the previously-`todo!()` enum arm in the
struct_enum example.

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01BRpeiPNoowJaMhAk2WYUMp

### CODE REVIEW

File: src/visitor.rs [hunk 1]
```diff
@@ -84,15 +107,26 @@ where
...
+        let state = self.state.take().expect("decoder state is always present");
+        let mut decoder = Decoder::with_state(state, one_shot::Wrapper::new(model));
+        let result = decoder.decode();
         let (_model, state) = decoder.into_inner();
         self.state = Some(state);
-        Ok(symbol)
+
+        // A one-shot wrapper always yields exactly one symbol; `None` here means
+        // the stream was exhausted before that symbol could be produced.
+        result?.ok_or(DecodeError::InvalidSymbol { symbol: 0 })
```
[the remark this answers is not in the document]
Comment (danieleades) [reply src/visitor.rs:135]: Confirmed empirically (`bool::decode_bytes(&[]) == Ok(false)`), but rejecting truncation outright isn't possible in an arithmetic coder: missing bits are indistinguishable from the implicit zero padding that *legitimate* streams end with — the decoder must zero-pad to decode valid messages at all, and nearly every byte string decodes to some syntactically valid value. 8bd8f87 corrects the docs that over-promised here (integrity requires outer framing/checksum). A stronger structural check lands with the upcoming size-report framework: under uniform weighting every value of a schema has one exact encoded length, so `decode_bytes` will reject any input whose byte length doesn't match the schema — catching all whole-byte truncations. Sub-byte corruption remains undetectable without redundancy, which is now documented.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: minnow-derive/src/parse.rs [hunk 2]
```diff
@@ -7,24 +7,94 @@ pub mod parse_struct;
...
+                return Err(format!(
+                    "float model lower bound ({min}) must not exceed the upper bound ({max})"
+                ));
+            }
+            let multiplier = 10_f64.powi(i32::from(*precision));
+            let steps = ((max - min) * multiplier).round();
+            #[allow(clippy::cast_precision_loss)]
+            let max_denominator = MAX_DENOMINATOR as f64;
+            // denominator = steps + 1
+            if !steps.is_finite() || steps + 1.0 > max_denominator {
```
[the remark this answers is not in the document]
Comment (danieleades) [reply minnow-derive/src/parse.rs]: Fixed in 8bd8f87: the macro-time check now mirrors `FloatModel::new` exactly — same f64 arithmetic, then the same integer comparison via `ToPrimitive::to_u128` — so accept/reject decisions agree at the boundary by construction. Regression tests pin the boundary cases (2^62 rejected, largest f64 below it accepted).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/impls.rs [hunk 3]
```diff
@@ -26,55 +26,66 @@ where
...
+impl Encodeable for OptionDiscriminant {
     fn encode<W>(&self, visitor: &mut EncodeVisitor<W>) -> io::Result<()>
     where
         W: BitWrite,
     {
         let value = match self {
-            Option__::Some => 0,
-            Option__::None => 1,
+            OptionDiscriminant::None => 0,
+            OptionDiscriminant::Some => 1,
```
[the remark this answers is not in the document]
Comment (danieleades) [reply src/impls.rs:71]: It's a prototype with no consumers. Not yet a concern

File: src/lib.rs [hunk 4]
```diff
@@ -6,20 +6,79 @@
...
+/// ```text
+/// D = 2^(PRECISION - 2) = 2^62 = 4_611_686_018_427_387_904
+/// ```
+///
+/// At `PRECISION = 64` the governing product `(2^P + 1) * D = 2^126 + 2^62`
+/// stays below `2^128 - 1`. Exceeding `D` risks silent round-trip corruption,
+/// so every model constructor validates its denominator against this bound —
+/// though at `~2^62` (over four quintillion distinguishable values) the bound
+/// is unreachable for any practical model.
+pub const MAX_DENOMINATOR: u128 = 1 << (PRECISION - 2);
```
[the remark this answers is not in the document]
Comment (danieleades) [reply src/lib.rs]: Correct — verified against the 0.4 source (`frequency_bits = ilog2 + 1`, `precision >= frequency_bits + 2`, and it's a `debug_assert` our `with_state` path bypasses). Fixed in d293d00: `MAX_DENOMINATOR = (1 << (PRECISION - 2)) - 1`, docs restate the upstream rule, and a unit test pins tightness (safe at the cap, unsafe one above), so the constants can't silently drift from the invariant again.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -8,70 +8,45 @@ name: Continuous integration
 jobs:
   test:
     name: test
-    runs-on: ${{ matrix.os }}
+    runs-on: ubuntu-latest
```
Comment (danieleades) [inline .github/workflows/CI.yml]: Why not test on other os targets?

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -8,70 +8,45 @@ name: Continuous integration
...
-          - build: nightly
-            os: ubuntu-latest
-            rust: nightly
-          # - build: macos
-          #   os: macos-latest
-          #   rust: stable
-          # - build: windows
-          #   os: windows-latest
-          #   rust: stable
+        rust: [stable, "1.85"]
```
Comment (danieleades) [inline .github/workflows/CI.yml]: I think this should be stable and nightly

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -8,70 +8,45 @@ name: Continuous integration
...
-      - uses: actions/checkout@v3
-      - uses: actions-rs/toolchain@v1
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@master
         with:
-          profile: minimal
           toolchain: ${{ matrix.rust }}
-          override: true
-      - uses: actions-rs/cargo@v1
+      - run: cargo test --workspace --all-features --all-targets
```
Comment (danieleades) [inline .github/workflows/CI.yml:25]: Why use a separate target for doing doc tests?

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -8,70 +8,45 @@ name: Continuous integration
...
+      - uses: dtolnay/rust-toolchain@stable
         with:
-          command: test
-          args: --all-features --workspace
+          components: clippy
+      - run: cargo clippy --workspace --all-targets --all-features
 
   fmt:
-    name: format
+    name: fmt
```
Comment (danieleades) [inline .github/workflows/CI.yml]: Redundant

File: .github/workflows/CI.yml [hunk 9]
```diff
@@ -8,70 +8,45 @@ name: Continuous integration
...
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions-rs/toolchain@v1
+      - uses: actions/checkout@v4
+      # `.rustfmt.toml` uses unstable options, so formatting requires nightly.
```
Comment (danieleades) [inline .github/workflows/CI.yml]: This comment not needed

File: .github/workflows/coverage.yml [hunk 10]
```diff
@@ -11,14 +11,15 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: checkout source
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
+      # Doctest coverage (--doctests) needs -Z unstable-options, so this job
```
Comment (danieleades) [inline .github/workflows/coverage.yml]: These verbose comments not useful

File: minnow-derive/src/parse/parse_enum.rs [hunk 11]
```diff
@@ -75,6 +75,7 @@ mod tests {
         }
         ; "tuple enum w model"
     )]
+    #[allow(clippy::needless_pass_by_value)]
```
Comment (danieleades) [inline minnow-derive/src/parse/parse_enum.rs]: Is this needed or can it be fixed properly?

File: minnow-derive/src/parse.rs [hunk 12]
```diff
@@ -102,8 +191,27 @@ mod tests {
         }
         ; "unit enum"
     )]
+    #[allow(clippy::needless_pass_by_value)]
```
Comment (danieleades) [inline minnow-derive/src/parse.rs]: As above

File: minnow-derive/src/write.rs [hunk 13]
```diff
@@ -11,6 +11,9 @@ pub fn write(receiver: Data) -> TokenStream {
     }
 }
 
+// The number of enum variants and their indices cannot realistically exceed
+// `u32::MAX`, so these `usize -> u32` casts cannot truncate in practice.
+#[allow(clippy::cast_possible_truncation)]
```
Comment (danieleades) [inline minnow-derive/src/write.rs]: Is there a better way to ensure this than just "it probably won't happen in practice"?

File: .github/workflows/CI.yml (hunk 7 above)
Comment (danieleades) [reply .github/workflows/CI.yml:25]: Cargo quirk rather than a choice: passing `--all-targets` *excludes* doctests (it expands to lib/bins/tests/benches/examples only), so a second `cargo test --doc` invocation is the standard way to keep doctests in CI. Same target dir, so the second run reuses the build.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/CI.yml (hunk 5 above)
Comment (danieleades) [reply .github/workflows/CI.yml]: Added in 4ece65a: macOS and Windows on stable, alongside the ubuntu stable/nightly legs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/CI.yml (hunk 6 above)
Comment (danieleades) [reply .github/workflows/CI.yml]: Done in 4ece65a — matrix is now `[stable, nightly]`. Note this drops the MSRV (1.85) leg from CI; `rust-version` stays declared in Cargo.toml but is no longer exercised. Shout if you'd rather keep a third leg for it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: minnow-derive/src/write.rs (hunk 13 above)
Comment (danieleades) [reply minnow-derive/src/write.rs]: Yes — fixed properly in 4ece65a: `u32::try_from(...).expect(...)` instead of `as` + allow. A panic in a proc macro surfaces as a compile error, so the absurd case fails loudly at build time instead of silently truncating the discriminant space.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: minnow-derive/src/parse/parse_enum.rs (hunk 11 above)
Comment (danieleades) [reply minnow-derive/src/parse/parse_enum.rs]: Fixed properly in 4ece65a (here and the same pattern in parse.rs): the lint fired because `syn::parse_str(&tokens.to_string())` only borrowed the tokens — `syn::parse2(tokens)` takes them by value, which is both the more direct API and removes the allow.

---
_Generated by [Claude Code](https://claude.ai/code)_
